### ISSUE TITLE
[Performance]: phi 3.5 vision model consuming high CPU RAM and the process getting killed

### ISSUE THREAD

**Issue by kuladeephx** [vllm-project/vllm#9190]
### Proposal to improve performance

I am trying to run phi3.5 vision instruct model with around 10k prompts. What I noticed with the increase in prompts my CPU RAM consumption keeps increasing and eventually the process gets killed. Its running fine for say small sample like 1000 prompts. My system configuration is 48 GB VRAM and 64GB CPU RAM. Noticed a similar pattern with PIXTRAL-12B-2409. Has anyone faced this issue?

I have tried the implementation by passing in batches of 1000 to llm.generate but still the CPU RAM keeps increasing
Below is the code implementation:
Ima using two images per prompt
from vllm import LLM, SamplingParams
llm = LLM(
        model="microsoft/Phi-3.5-vision-instruct",
        gpu_memory_utilization=0.7,
        trust_remote_code=True,
        max_model_len=4096,
        limit_mm_per_prompt={"image": 4},
    )
sampling_params = SamplingParams(max_tokens=100, temperature=0.0)
outputs = llm.generate(prompt_list, sampling_params=sampling_params)

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
It was fixed just earlier today by #9160. Can you try pulling the latest `main` branch?

**Comment by kuladeephx** [comment]
I have tried with the latest main branch, its still the same

**Comment by DarkLight1337** [comment]
cc @Isotr0py

**Comment by Isotr0py** [comment]
@kuladeephx I noticed you didn't set `max_num_seqs` when initialized `LLM`. Can you check if setting `max_num_seqs=2` would solve this?

**Comment by kuladeephx** [comment]
Have tried now with max_num_seqs=2, still the same

**Comment by Isotr0py** [comment]
Does this issue only encounter on `phi-3.5-vision` or encounter on other VLMs? I'm reproducing this issue with below commands, but it would cost some time for me to run 10k prompts on T4:
```shell
python examples/offline_inference_vision_language.py -m phi3_v --num-prompts 10000
```

**Comment by kuladeephx** [comment]
Thanks for the try, so the above trail does it apply 10000 prompts with no repeated images? 
yes noticed similar pattern with mistralai/Pixtral-12B-2409

**Comment by Isotr0py** [comment]
Can you check if other models like `Qwen/Qwen2-VL-2B-Instruct` and `Qwen/Qwen2-VL-7B-Instruct` also have this issue?

I'm not sure if this issue relates to composite weight loading, we can have a check on Qwen2-VL models since it doesn't use composite weight loading.

**Comment by kuladeephx** [comment]
Sure will try, was there any reference previously of people trying these above mentioned models with say 10000 prompts or higher. Just curious if anyone has tried in a similar scale

**Comment by kuladeephx** [comment]
@Isotr0py, can you pls let me know, were you able to reproduce the issue with 10000 prompts

**Comment by Isotr0py** [comment]
No, I can't reproduce this issue with 10k prompts. I didn't observe significant CPU RAM usage increase during the inference. In fact, my CPU RAM usage is mostly about 25% (about 4GB on my 16GB RAM device) all the time after loading the model.

Can you provide a full script that can reproduce this issue?

**Comment by Isotr0py** [comment]
The memory leak issue is a little bit similar to #8629 since it also occurred after long running. 

Can you check if the solution provided by https://github.com/vllm-project/vllm/issues/8629#issuecomment-2363312483 work?

**Comment by kuladeephx** [comment]
@Isotr0py , I have tried a similar implementation of clearing GPU cache and garbage collection earlier, but it doesn't solve the issue. I am attaching the script and sample data(dummy_image_data.zip), you can try the experiment with total_no_of_prompts = 300 vs total_no_of_prompts = 600 and see the difference. Additional packages which might be required are requests, memory_profiler. I am attaching memory profile screenshots for  total_no_of_prompts = 300 vs 600 case

`import requests
import pandas as pd
import gc
import torch
from memory_profiler import profile
from tqdm import tqdm
from PIL import Image
from vllm import LLM, SamplingParams
from huggingface_hub import login, logout

hf_key = ''





def break_list_into_chunks(full_list, chunk_size):
    chunked_list = []

    # Iterate over the input list and add chunks to the result list
    for i in range(0, len(full_list), chunk_size):
        chunked_list.append(full_list[i:i + chunk_size])

    return chunked_list



@profile
def run_multimodal_model(prompt_list, llm_batch_size):
    login(token=hf_key)
    llm = LLM(
        model="microsoft/Phi-3.5-vision-instruct",
        gpu_memory_utilization=0.7,
        trust_remote_code=True,
        max_model_len=4096,
        max_num_seqs=2,
        limit_mm_per_prompt={"image": 4},
        
    )
    sampling_params = SamplingParams(max_tokens=100, temperature=0.0)
    final_out_list = []
    # Breaking prompts into chunks
    prompt_chunks = break_list_into_chunks(prompt_list, llm_batch_size)
    
    del prompt_list
    for prompt_chunk in prompt_chunks:
        
        outputs = llm.generate(prompt_chunk, sampling_params=sampling_params)
        torch.cuda.empty_cache()
        gc.collect()
        

        outputs = [i.outputs[0].text.strip() for i in outputs]
        final_out_list.extend(outputs)
        del outputs
        del prompt_chunk

    
    
    llm = None
    
    del llm

    # Clear the GPU cache
    torch.cuda.empty_cache()

    return final_out_list



def generate_prompt(image_url):
    try:
        # Reading the image from url
        image_val = Image.open(requests.get(image_url, stream=True).raw)
        image1 = image_val
        image2 = image_val
    except:
        return -1
    
    prompt_val = f"""
<|user|>
<|image_1|>
<|image_2|>
Are these two image same? Just answer yes or no.
"""

    final_prompt = {
        "prompt": prompt_val,
        "multi_modal_data": {"image": [image1, image2]},
    }

    return final_prompt


@profile
def get_final_output(df, total_no_of_prompts, llm_batch_size):
    unique_image_urls = df["url"].unique().tolist()
    #Subsetting number of prompts
    unique_image_urls = sorted(unique_image_urls)[:total_no_of_prompts]
    
    print("---------------Generating Pompts---------------")
    prompts_list = [generate_prompt(image_url) for image_url in tqdm(unique_image_urls)]
    prompts_list = [x for x in prompts_list if x!=-1]
    
    final_outputs = run_multimodal_model(prompts_list,llm_batch_size)
    return final_outputs


if __name__ == "__main__":

    input_df = pd.read_csv("dummy_image_data.zip")
    
    
    result = get_final_output(df =input_df,
                              total_no_of_prompts = 300,
                              llm_batch_size = 150)
    
    """
    df: input dataframe
    total_no_of_prompts: total number of prompts to run
    llm_batch_size: no of prompts to be executed by LLM at once
    """`
[dummy_image_data.zip](https://github.com/user-attachments/files/17324120/dummy_image_data.zip)
<img width="672" alt="300_prompts" src="https://github.com/user-attachments/assets/6a873693-ed2f-4eaf-925b-33bba6564d3e">
<img width="679" alt="600_prompts" src="https://github.com/user-attachments/assets/ceddcbaa-a2cd-42b3-822f-80d652ef6106">

**Comment by Isotr0py** [comment]
Thanks, I can reproduce this issue with the provided script. Let me use vLLM's profiler to see what's wrong.

**Comment by kuladeephx** [comment]
@Isotr0py, any lead on the issue?

**Comment by Isotr0py** [comment]
I guess it's the Phi-3-Vision's image embedding causing the memory leak. I observed a high memory allocation about `aten::cat` (which is widely used in phi3v image embedding as well):
![20241011131111](https://github.com/user-attachments/assets/a6617059-4e0f-41a5-8216-c833317ee096)

**Comment by kuladeephx** [comment]
Have noticed a similar memory increase with pixtral-12B-2409 as well, so wondering if this has to do with vllm

**Comment by Isotr0py** [comment]
I also profiled the `phi3_v` forwarding, seems that the memory leak not related to the model forwarding (the memory increasement from model forwarding only occurs on first time prefill). 

The profiling log: [phi3v_forward_profile.log](https://github.com/user-attachments/files/17340311/phi3v_forward_profile.log)

I suspect the source of memory leak comes from other parts of `llm_engine` instead :(

**Comment by kuladeephx** [comment]
Thanks for analysis, In that case this seems to affect multimodal models? Do you think the issue can be identified?

And what's the difference between LLM.generate and LLM.chat (excluding the use case of usage as a chat model), I tried the same set with LLM.chat and noticed memory consumption is less compared using LLM.generate

**Comment by Isotr0py** [comment]
At least in my test, other multimodal models like InternVL2 also encountered this issue. For example, here is the memory profile with `InternVL2-1B`:
![internvl](https://github.com/user-attachments/assets/10f6e9fd-81ea-4b2e-897c-2e0018031b5f)


Since there is no memory leak from the model forwarding, the source of leaking memory might be various (`scheduler` or `block_manager` etc). I'm afraid that I can't help with this, because I'm not very familiar with these implementations.

> And what's the difference between LLM.generate and LLM.chat (excluding the use case of usage as a chat model), I tried the same set with LLM.chat and noticed memory consumption is less compared using LLM.generate

This is interesting, because LLM.chat just applies chat_template before calling LLM.generate.

**Comment by kuladeephx** [comment]
Thanks, do you think is there anyone else who can help with this issue?

**Comment by DarkLight1337** [comment]
@WoosukKwon @youkaichao do you have any tips on how to debug memory leaks in this part?

**Comment by youkaichao** [comment]
CPU side memory analysis is more complicated than GPU :(

**Comment by ShiningMaker** [comment]
Hi, is there any update on this? I encountered the same issue with the GPTQ model. After setting the utilization to 0.95 on a 40G GPU and running a few datasets, I experienced an out-of-memory error.

**Comment by DarkLight1337** [comment]
I am not sure how relevant this issue is in the latest version of vLLM. #14805 has resolved a known issue about input cache but this wasn't implemented when this issue was first created.

**Comment by kuladeephx** [comment]
@DarkLight1337, have tried with latest main branch, still the same issue

**Comment by DarkLight1337** [comment]
Can you try `--disable-mm-preprocessor-cache` and see if it improves the RAM usage?

**Comment by DarkLight1337** [comment]
Can you try to reduce the `limit_mm_per_prompt`?

**Comment by kuladeephx** [comment]
tried with limit_mm_per_prompt=2, still the same

**Comment by Somasundaram-Palaniappan** [comment]
This is till an issue guys. memory increases non stop on the latest release. 

![Image](https://github.com/user-attachments/assets/72f6db33-191f-455e-8299-d66eca8ee0ce)

**Comment by DarkLight1337** [comment]
> This is till an issue guys. memory increases non stop on the latest release.
> 
> ![Image](https://github.com/user-attachments/assets/72f6db33-191f-455e-8299-d66eca8ee0ce)

Try passing `--disable-mm-processor-cache`

**Comment by DarkLight1337** [comment]
But even if you don't disable it, the memory usage should eventually level off.

**Comment by YueChenkkk** [comment]
Same issue with vllm 0.7.2, CPU RAM keeps increasing when serving Qwen2.5-7B. The problem persists with or without `--enable_prefix_caching`

**Comment by scpljx** [comment]
same problem when I trying qwen2.57bvl/qwen2.57b, it seems like only the inferences is done, memory can release, memory will increasing until the whole inference is down

Line    Mem usage    Increment  Occurrences   Line Contents
=============================================================
          13184.0 MiB   2250.5 MiB           7           outputs = llm.chat(conversation_batch, sampling_params=sample_params)
          13184.0 MiB      0.0 MiB           7                   if(idx > 5): break

Line    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    64  11551.8 MiB -1310719.5 MiB        1001       for idx, conversation_batch in enumerate(tqdm_bar):
    65  11551.8 MiB -1308962.3 MiB        1000           request_id = f"req_{idx}_{time.time()}" 
    66                                                 # conversation_batch 现在是 List[List[ChatCompletionMessageParam]]
    67  11551.8 MiB -1309832.9 MiB        1000           outputs = llm.chat(conversation_batch, sampling_params=sample_params)
    68                                                 # if(idx > 5): break

<img width="1260" height="540" alt="Image" src="https://github.com/user-attachments/assets/a5ac26f7-b392-42bd-a9f6-da939274b474" />

**Comment by DarkLight1337** [comment]
vLLM has a 8 GB cache (on CPU) for storing the HF processor outputs, so it is normal for the memory to increase a bit initially. However it should eventually stabilize once the cache is filled.

**Comment by scpljx** [comment]
> > This is till an issue guys. memory increases non stop on the latest release.
> 
> Try passing `--disable-mm-preprocessor-cache`


Thank you, I resolve this problem by passing `--disable-mm-preprocessor-cache`. While if let it default value, the memory will keep increasing, the more input token, the more memory vllm occupy. The memory usuage will not stable until the whole inference is done.

for idx, conversation_batch in enumerate(tqdm_bar):
        request_id = f"req_{idx}_{time.time()}" 
        # conversation_batch 现在是 List[List[ChatCompletionMessageParam]]
        outputs = llm.chat(conversation_batch, sampling_params=sample_params)
        # if(idx > 10): break

<img width="1260" height="540" alt="Image" src="https://github.com/user-attachments/assets/95814db6-7c53-4fc6-8090-c6f58a9d4215" />
